files/routes/comments.py

```diff
@@ -364,15 +340,68 @@
 	if v.marseyawarded and parent_post.id not in ADMINISTRATORS and marseyaward_body_regex.search(body_html):
 		return {"error":"You can only type marseys!"}, 403
 
-	parent_post.comment_count += 1
-	g.db.add(parent_post)
-
 	g.db.commit()
 
 	if request.headers.get("Authorization"): return c.json
 	return {"comment": render_template("comments.html", v=v, comments=[c], ajax=True)}
 
 
+def comment_on_publish(comment):
+	"""
+	Run when comment becomes visible: immediately for non-filtered comments,
+	or on approval for previously filtered comments.
+	Should be used to update stateful counters, notifications, etc. that
+	reflect the comments users will actually see.
+	"""
+	# TODO: Get this out of the routes and into a model eventually...
+
+	# Shadowbanned users are invisible. This may lead to inconsistencies if
+	# a user comments while shadowed and is later unshadowed. (TODO?)
+	if comment.author.shadowbanned:
+		return
+
+	# Comment instances used for purposes other than actual comments (notifs,
+	# DMs) shouldn't be considered published.
+	if not comment.parent_submission:
+		return
+
+	# Generate notifs for: mentions, post subscribers, parent post/comment
+	to_notify = NOTIFY_USERS(comment.body, comment.author)
+
+	post_subscribers = g.db.query(Subscription.user_id).filter(
+			Subscription.submission_id == comment.parent_submission,
+			Subscription.user_id != comment.author_id,
+		).all()
+	to_notify.update([x[0] for x in post_subscribers])
+
+	parent = comment.parent
+	if parent and parent.author_id != comment.author_id:
+		to_notify.add(parent.author_id)
+
+	for uid in to_notify:
+		notif = Notification(comment_id=comment.id, user_id=uid)
+		g.db.add(notif)
+
+	# Comment counter for parent submission
+	comment.post.comment_count += 1
+	g.db.add(comment.post)
+
+	# Comment counter for author's profile
+	comment.author.comment_count = g.db.query(Comment).filter(
+			Comment.author_id == comment.author_id,
+			Comment.parent_submission != None,
+			Comment.is_banned == False,
+			Comment.deleted_utc == 0,
+		).count()
+	g.db.add(comment.author)
+
+	# Generate push notifications if enabled.
+	if PUSHER_ID != 'blahblahblah' and comment.author_id != parent.author_id:
+		try:
+			gevent.spawn(pusher_thread, f'{request.host}{parent.author.id}',
+				comment, comment.author_name)
+		except: pass
```

Comment:
Aside from this, this looks fine to me!

When do we expect this to trigger? I'm a bit worried about silent-failure stuff happening.

Reply:
For this particular code scanning flag? This is the logic that does push notifications to a user's browser. The entire feature may well even be disabled (by the branch on L399) unless you have the `PUSHER_ID` env variable set on prod.

For what it's worth, there's no regression here at least: it just thinks it's new code because I moved it. Also purely harmless if it fails silently: users will still get notified in their site inbox, just not with their browser. We perhaps ought to refactor the push notif logic at some point in the future anyway, though.